pkg/handlers/handler_base.go

```diff
@@ -0,0 +1,171 @@
+	return errorResponse{
+		Errors:  []errorItem{},
+		Code:    code,
+		Message: message,
+		TraceID: traceID,
+	}
+}
+
+func (r *errorResponse) withMap(errMap map[string]string) {
+	for k, v := range errMap {
+		r.Errors = append(r.Errors, errorItem{
+			Field:   k,
+			Message: v,
+		})
+	}
+}
+
+// WriteErrorResponse writes a response for a given application error
+func (b HandlerBase) WriteErrorResponse(ctx context.Context, w http.ResponseWriter, appErr error) {
+	logger, ok := appcontext.Logger(ctx)
+	if !ok {
+		logger = b.logger
+	}
+
+	traceID, ok := appcontext.Trace(ctx)
+	if !ok {
+		traceID = uuid.New()
+		logger.With(zap.String("traceID", traceID.String()))
+	}
+
+	// get code and response
+	var code int
+	var response errorResponse
+	switch appErr := appErr.(type) {
+	case *apperrors.UnauthorizedError:
+		// 4XX errors are not logged as errors, but are for client
+		logger.Info("Returning unauthorized response from handler", zap.Error(appErr))
+		code = http.StatusUnauthorized
+		response = newErrorResponse(
+			http.StatusUnauthorized,
+			"Unauthorized",
+			traceID,
+		)
+	case *apperrors.QueryError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.NotificationError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Failed to send notification",
+			traceID,
+		)
+	case *apperrors.ExternalAPIError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.ContextError:
+		logger.Error("Returning server error response from handler", zap.Error(appErr))
+		code = http.StatusInternalServerError
+		response = newErrorResponse(
+			http.StatusInternalServerError,
+			"Something went wrong",
+			traceID,
+		)
+	case *apperrors.ValidationError:
+		logger.Info("Returning bad request error from handler", zap.Error(appErr))
+		code = http.StatusBadRequest
+		response = newErrorResponse(
+			http.StatusBadRequest,
+			"Bad request",
+			traceID,
+		)
+		response.withMap(appErr.Validations.Map())
+	case *apperrors.MethodNotAllowedError:
+		logger.Info("Returning method not allowed error from handler", zap.Error(appErr))
+		code = http.StatusMethodNotAllowed
+		response = newErrorResponse(
+			http.StatusMethodNotAllowed,
+			"Method not allowed",
+			traceID,
+		)
+	case *apperrors.ResourceConflictError:
+		logger.Info("Returning resource conflict error from handler", zap.Error(appErr))
+		code = http.StatusConflict
+		response = newErrorResponse(
+			http.StatusConflict,
```

Comment:
I think where we've been using ResourceConflict, we really want to return 422 Unprocessable Entity instead.

Reply:
I think the way we're using it is a 409. ie. the current resources is in a state (submitted) that conflicts with the new one (attempting to resubmit). If it's not a 409, we should probably be renaming that error.

I think 422 would be if someone had `status: "weirdstatus"`, but the json is correct. Which might be something to visit compared to the `Bad Requests` we're defaulting to for any validation error.

Reply:
Yes and no. 409 suggests there's a conflict, but in the way that db table is locked or resource is currently being updated. A 422 says that the json is correct and as it is, it can't be executed. Here's the definition of 422 from the RFC: https://tools.ietf.org/html/rfc4918#section-11.2. That said, I do know that the more important thing here is surfacing the right error message so that we can surface actionable errors for the client and the user, so my comment is not blocking.